pycgm/static.py

```diff
@@ -0,0 +1,299 @@
+    >>> rtoe = np.array([442.82, 381.62, 42.66])
+    >>> ltoe = np.array([39.44, 382.45, 41.79])
+    >>> static_info = [[0.03, 0.15, 0],
+    ...                [0.01, 0.02, 0]]
+    >>> ankle_joint_center = [np.array([393.76, 247.68, 87.74]),
+    ...                       np.array([98.75, 219.47, 80.63]),
+    ...                       [[np.array([394.48, 248.37, 87.72]),
+    ...                         np.array([393.07, 248.39, 87.62]),
+    ...                         np.array([393.69, 247.78, 88.73])],
+    ...                        [np.array([98.47, 220.43, 80.53]),
+    ...                         np.array([97.79, 219.21, 80.76]),
+    ...                         np.array([98.85, 219.60, 81.62])]]]
+    >>> delta = 0
+    >>> [np.around(arr,2) for arr in """
+    """foot_joint_center(rtoe, ltoe,static_info,ankle_joint_center)] """
+    """#doctest: +NORMALIZE_WHITESPACE
+    [array([442.82, 381.62,  42.66]), array([ 39.44, 382.45,  41.79]), """
+    """array([[[442.89, 381.76,  43.65],
+            [441.89, 382.  ,  42.67],
+            [442.45, 380.7 ,  42.82]],
+           [[ 39.51, 382.68,  42.76],
+            [ 38.5 , 382.15,  41.93],
+            [ 39.76, 381.53,  41.99]]])]
+    """)
+    # REQUIRED MARKERS:
+    # RTOE
+    # LTOE
+    toes = [rtoe, ltoe]
+
+    # REQUIRE JOINT CENTER & AXIS
+    # KNEE JOINT CENTER
+    # ANKLE JOINT CENTER
+    # ANKLE FLEXION AXIS
+
+    # Dealing with the Incorrect Axis
+    foot_axis = []
+    for side in range(2):  # calculate the left and right foot axes
+        # Z_axis taken from toe marker to Ankle Joint Center; then normalized.
+        z_axis = np.array(ankle_joint_center[side]) - np.array(toes[side])
+        with np.errstate(invalid='ignore'):
+            z_axis = np.divide(z_axis, np.nan_to_num(np.linalg.norm(z_axis)))
+
+        # Calculated flexion axis from Ankle Joint Center; then normalized.
+        flexion_axis = (np.array(ankle_joint_center[2][side][1])
+                        - np.array(ankle_joint_center[side]))
+        with np.errstate(invalid='ignore'):
+            flexion_axis = np.divide(flexion_axis,
+                                     np.nan_to_num(np.linalg.norm(flexion_axis)))
+
+        # X_axis taken from cross product of Z_axis and the flexion axis.
+        x_axis = np.cross(flexion_axis, z_axis)
+        with np.errstate(invalid='ignore'):
+            x_axis = np.divide(x_axis, np.nan_to_num(np.linalg.norm(x_axis)))
+
+        # Y_axis take from cross product of Z_axis and X_axis (perpendicular).
+        # Then normalized.
+        y_axis = np.cross(z_axis, x_axis)
+        with np.errstate(invalid='ignore'):
+            y_axis = np.divide(y_axis, np.nan_to_num(np.linalg.norm(y_axis)))
+
+        # Apply static offset angle to the incorrect foot axes
+        # static offset angle are taken from static_info variable in radians.
+
+        alpha = np.radians(np.around(np.degrees(static_info[side][0]), 5))
```

Comment:
why is this rounding?

Reply:
Actually it's weirder than that, this code takes values that are already in radians, then converts them to degrees and back to radians. I tried using the original radian values but because the conversion process creates some trailing decimals the results are off from the test values.

Should I go ahead and use the original radian values (skip the conversion back and forth) and adjust the test values accordingly?
Or is there a historic reason for the process?

As for the rounding, much as above taking it off led to conflicts with the test values.

Reply:
Lets leave a comment that it used to do this, but fix it.  I think there may have historically been an issue of rounding because we suspected a float and not a double being used in the comparison, but not sure.